src/ReverseProxy/Service/DynamicEndpoint/ReverseProxyConventionBuilder.cs

```diff
@@ -0,0 +1,110 @@
+        /// </summary>
+        /// <param name="convention">The convention to add to the builder.</param>
+        public void Add(Action<EndpointBuilder> convention)
+        {
+            _ = convention ?? throw new ArgumentNullException(nameof(convention));
+
+            _conventions.Add(convention);
+        }
+
+        /// <summary>
+        /// Configures the endpoints for all routes 
+        /// </summary>
+        /// <param name="convention">The convention to add to the builder.</param>
+        /// <returns></returns>
+        public ReverseProxyConventionBuilder Configure(Action<IEndpointConventionBuilder> convention)
+        {
+            _ = convention ?? throw new ArgumentNullException(nameof(convention));
+
+            void Action(EndpointBuilder endpointBuilder)
+            {
+                var conventionBuilder = new EndpointBuilderConventionBuilder(endpointBuilder);
+                convention(conventionBuilder);
+            }
+
+            Add(Action);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the endpoints for all routes 
+        /// </summary>
+        /// <param name="convention">The convention to add to the builder.</param>
+        /// <returns></returns>
+        public ReverseProxyConventionBuilder Configure(Action<IEndpointConventionBuilder, ProxyRoute> convention)
+        {
+            _ = convention ?? throw new ArgumentNullException(nameof(convention));
+
+            void Action(EndpointBuilder endpointBuilder)
+            {
+                var routeConfig = endpointBuilder.Metadata.OfType<RouteConfig>().Single();
+                var conventionBuilder = new EndpointBuilderConventionBuilder(endpointBuilder);
+                convention(conventionBuilder, routeConfig.ProxyRoute.DeepClone());
+            }
+
+            Add(Action);
+
+            return this;
+        }
+
+        /// <summary>
+        /// Configures the endpoints for all routes 
+        /// </summary>
+        /// <param name="convention">The convention to add to the builder.</param>
+        /// <returns></returns>
+        public ReverseProxyConventionBuilder Configure(Action<IEndpointConventionBuilder, ProxyRoute, Cluster> convention)
+        {
+            _ = convention ?? throw new ArgumentNullException(nameof(convention));
+
+            void Action(EndpointBuilder endpointBuilder)
+            {
+                var routeConfig = endpointBuilder.Metadata.OfType<RouteConfig>().Single();
+
+                var cluster = routeConfig.Cluster?.Config.Cluster;
+                var proxyRoute = routeConfig.ProxyRoute;
+                var conventionBuilder = new EndpointBuilderConventionBuilder(endpointBuilder);
+                convention(conventionBuilder, proxyRoute.DeepClone(), cluster?.DeepClone());
+            }
+
+            Add(Action);
+
+            return this;
+        }
+
+        internal class EndpointBuilderConventionBuilder : IEndpointConventionBuilder
```

Comment:
```suggestion
        private class EndpointBuilderConventionBuilder : IEndpointConventionBuilder
```

Type used anywhere else?